### PR TITLE
Supporting line wise yanks, including paste and undo.

### PR DESCRIPTION
**Pull request by springcomp**
This is my take at [#333](https://github.com/lzybkr/PSReadLine/issues/333).

I thought best to create a `ViRegister` class to replace the current `_clipboard` `String` because I wanted to support pasting "line wise" irrespective of where the cursor is at when pasting occurs. I created an `event` that gets raised each time the buffer is about to change as part of a paste operation, to hook into undo/redo without too much coupling.

Please, let me know your thoughts and suggestions for improvements.

### CODE REVIEW

**Comment by springcomp** [comment]
I updated this pull request via `force push` using the accepted answer from [this question](https://stackoverflow.com/questions/13561004/merging-upstream-git-changes-with-pending-pull-requests).

File: test/StringBuilderLinewiseExtensionsTests.cs
Comment (lzybkr) [inline test/StringBuilderLinewiseExtensionsTests.cs]: How is this different from `StringBuilderLinewiseExtensions_LinewiseYank_Fragment`?

File: PSReadLine/ViRegister.cs
Comment (lzybkr) [inline PSReadLine/ViRegister.cs]: Does this really need to be public?

File: PSReadLine/ViRegister.cs
Comment (lzybkr) [inline PSReadLine/ViRegister.cs]: Does this really need to be public? (I see it used in a test, but it becomes a supported API if it's public.)

**Review by lzybkr — CHANGES_REQUESTED** [review]
Sorry it took forever for me to take a look
I have some general comments that need to be addressed and I'll try to take a closer look then, or maybe @daxian-dbw will beat me to it.

File: PSReadLine/ViRegister.cs
Comment (springcomp) [reply PSReadLine/ViRegister.cs]: You're right. I will make sure all occurrences use the `Linewise` form.

File: PSReadLine/ViRegister.cs
Comment (springcomp) [reply PSReadLine/ViRegister.cs]: Sure, that would be more compact.
I will do that.

File: test/StringBuilderLinewiseExtensionsTests.cs
Comment (springcomp) [reply test/StringBuilderLinewiseExtensionsTests.cs]: You are right, it is not.
I will keep only one.

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: This goes hand in hand with why `StringBuilderLinewiseExtensions` should be made public. In essence, this does not _need_ to be public. It’s just that I am using TDD to implement my developments and I do not see an elegant way to test `StringBuilderLinewiseExtensions` without it being public.

That said, the public / external _behavior_ of the code is ultimately unit-tested as well.

Would you be OK if I added some `#if DEBUG` conditionals so that this code is tested during debug builds only ? Keeping in mind that the purpose of this code is to have a few public facing API that are always tested anyway (in my case, by new tests in the `YankPasteTest.VI.cs` file).

Another solution would be to use a dummy .snk strong-key to sign both assemblies and make use of the `InternalsVisibleTo` attributes.

As a third solution, I would make those classes non-public as appropriate and remove the corresponding unit-tests.

Can you weigh in on this issue ? Alternatively, I can ask a question as a separate issue to have a broader consensus involving more developers you may trust.

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: I will deal with this class in accordance to [this comment](https://github.com/PowerShell/PSReadLine/pull/811#discussion_r336831934)

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: That was just a matter of taste, but I understand that making the enclosing class public is an important consideration. I will change the class’ visibility and name, and not make this an extension method.

File: PSReadLine/ViRegister.cs
Comment (springcomp) [reply PSReadLine/ViRegister.cs]: It does not, no. I will make it internal.

Comment (springcomp) [reply PSReadLine/ViRegister.cs]: Well, I need some sort of callback mechanism between the `ViRegister` class and the Undo/Redo feature of the `PSReadLine` class. But it certainly does not need to be public.

If you prefer, I can change this into a method that registers a callback instead of an event.

File: PSReadLine/ViRegister.cs
Comment (springcomp) [reply PSReadLine/ViRegister.cs]: It does not, I will change this to internal, or according to decisions made as per [this comment](https://github.com/PowerShell/PSReadLine/pull/811#discussion_r337091480).

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: Sure, I can.

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: This method supports a scenario whereby the user specifies a number of lines to be taken in consideration for yanking.

The `lineCount` argument must not be 0, but is otherwise unbounded. If the user specifies more lines than there are available, making the requested range of lines larger than what it can be, I still want this method to return a valid range of offsets into the buffer.
For instance, if you go to the next to last line of a file in vim and ask <kbd>42</kbd>, <kbd>yy</kbd>, this would nevertheless successfully yank two lines of text in the register.

So I could add a check that `lineCount` is greater than 0.

The `lineIndex` argument comes from the cursor position via a call to the `GetLogicalLineNumber()` so is guaranteed to be valid, and thus, this method safe to return a valid range of offsets into the buffer.

Are you OK for an extra check on `lineCount` at the start of the method, an extra `Debug.Assert(true == startPositionIdentified)` just before returning the range and more comments in the method documentation?

Comment (lzybkr) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: As a general purpose method (extension to StringBuilder), it seems wrong.

As a special purpose method with good comments explaining such boundary conditions, it's fine.

I would make sure `lineCount == 0` is handled reasonably - e.g. user types <kbd>0yy</kbd> or something like that. We don't want exceptions in release builds, so assertions might not be enough.

File: PSReadLine/ViRegister.cs
Comment (lzybkr) [reply PSReadLine/ViRegister.cs]: You have a single handler and it looks like it could be inlined. Do you have other callbacks in mind?

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (lzybkr) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: I'll let @daxian-dbw chime in here. In the past we used `InternalsVisibleTo` but that may be less desirable now.

Another alternative is for a "test only" build where the unit tests are included in the same binary.

Private reflection is yet another (probably wrong) option.

File: PSReadLine/ViRegister.cs
Comment (springcomp) [reply PSReadLine/ViRegister.cs]: I don’t have other callbacks in mind, no. But, I fail to understand what you are driving at.

I wanted to avoid a direct coupling between the `ViRegister` and the `PSReadLine` class. In essence, the `ViRegister` needs to call a method on the `PSReadLine` to undo a paste operation.

I could achieve this by passing a reference to the `PSReadLine`  `_singleton` object as an argument to the `ViRegister` constructor, but that is what I was trying to avoid.

The only two solutions I see are :
a) An event.
b) A method that accepts a callback.

Sorry about my misunderstanding.

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: Sure. Thanks for that.

File: PSReadLine/ViRegister.cs
Comment (lzybkr) [reply PSReadLine/ViRegister.cs]: The `_singleton` object is pretty well baked in at this point, so a callback here felt like slight over-engineering.

File: PSReadLine/StringBuilderLinewiseExtensions.cs
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: I like the idea to include the tests in the same assembly with conditional compilation.

File: PSReadLine/ViRegister.cs
Comment (springcomp) [reply PSReadLine/ViRegister.cs]: I have made the `_clipboard` member static and initialized both `_singleton` and `_clipboard` objects from a static constructor.

**Comment by springcomp** [comment]
Sorry, I force-pushed again because there was structural changes to address your general comments.
You can now take a closer look.

From now on I will address your feedback with additional commits.

**Comment by springcomp** [comment]
This PR still shows changes requested even though I have carried out the changes.
@lzybkr  can you please re-review?

**Review by lzybkr — APPROVED** [review]

**Comment by lzybkr** [comment]
It looks like not all of the new tests were run in CI - presumably because test discovery isn't searching Microsoft.PowerShell.PSReadLine2.dll.

I'd like @daxian-dbw to approve including tests like this before merging.

**Comment by springcomp** [comment]
The latest commit adds discovery and execution of thoses unit-tests.

For this purpose, the main module is compiled with an alternate set of options and to a target assembly named `Microsoft.PowerShell.PSReadLine2.UnitTests.dll` so that it does not conflict with the release assembly.

@daxian-dbw, are you OK with these changes?

**Comment by daxian-dbw** [comment]
Can we move the test code to `PSReadLine.Tests` (under `PSReadLine\test`)? We can make `Microsoft.PowerShell.PSReadLine2` internal visible to `PSReadLine.Tests`.

**Comment by springcomp** [comment]
I thought about doing this but was deterred as per @lzybkr’s [comment](https://github.com/PowerShell/PSReadLine/pull/811#discussion_r337137904).

I would be happy to go ahead, but I do not know how to handle `InternalsVisibleTo` and assembly delay-signing?

For this to work, there would need to be a strong-name snk file in the repository for signing both assemblies. The way I handled this in the past was for the owner of the repository to include an encrypted version of the release strong-name snk file in the repository, and use AppVeyor secure strings to decrypt the file during the build.

Would that work for you @daxian-dbw?

**Comment by daxian-dbw** [comment]
I don't think `InternalsVisibleTo` requires the involved assemblies to be strong-name signed. Quoted the following from [the doc of this attribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=netframework-4.8#remarks)

> If both assemblies are unsigned, the assemblyName argument consists of the name of the friend assembly, specified without a directory path or file name extension.

I think you can just add `[assembly: InternalsVisibleTo("PSReadLine.Tests")]` to `PSReadLine\PSReadLine\AssemblyInfo.cs` and it should work.

**Comment by springcomp** [comment]
I can’t believe I just learned about that fact today 😱😱!

**Comment by daxian-dbw** [comment]
Thanks @springcomp!
I made a minor change to move the `InternalsVisibleTo` declaration to `AssemblyInfo.cs`, where other assembly scope attributes are put. That change breaks the build and it seems that file gets excluded somehow. I will fix it.

**Comment by daxian-dbw** [comment]
Build fixed. @springcomp Could you please review my changes? Thanks!

**Review by springcomp — COMMENTED** [review]
Yes, that is the reason why I created a separate file but this solution is certainly better.

**Comment by springcomp** [comment]
With this PR, I intend to work next on implementing <kbd>dd</kbd> (delete line) and, possibly <kbd>S</kbd> (substitute line) as natural extensions, so am looking forward to seeing this merged.

**Pull request merged**
